lib/train-habitat/httpgateway.rb

```diff
@@ -15,6 +17,34 @@ def initialize(host)
       def services
         JSON.parse(Net::HTTP.get_response(uri).body)
       end
+
+      def service(origin, name)
+        selected = services.select(&by_origin(origin))
+                           .select(&by_name(name))
+
+        validate!(selected, origin, name)
+
+        selected.first
+      end
```

Comment:
Bugs me how procedural this is.
```ruby
def service(origin, name)
  selected = services.select(&by_origin(origin)).select(&by_name(name))

  selected.first
ensure
  validate!(selected)
end
```
You might pull out separate validation statements to clarify the logic here, and at least from my perspective even out the abstraction level.
```ruby
def service(origin, name)
  selected = services.select(&by_origin(origin)).select(&by_name(name))

  selected.first
ensure
  raise NoServicesFoundError.new(origin, name) if selected.empty?
  raise MultipleServicesFoundError.new(origin, name) if selected.size > 1
end
```

Reply:
When is `ensure` evaluated? If we have no services selected and try to do `selected.first` we may throw an error there.